package chap22_IO;

import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.io.Reader;
import java.io.Writer;

public class _08_byteStreamToCharStream {

	public static void main(String[] args) {
		// TODO Auto-generated method stub
		String str = "";
		write("aaa");
		String str = read();
		System.out.println(str);
		
		
	}
	//try catch와 throws의 차이점
	//try catch는 exception에 대한 처리까지 가능한 기능
	// throws는 메소드를 호출한 부분에 에러가 나면 에러를 던지겠다. 알아서 처리해달라
	
	public static void write(String str) {
		try {
			OutputStream os = new FileOutputStream("C:/java/hp01.txt");
			
			//OutputStreamWriter을 이용해서 
			//바이트스트림 -> 문자스트림으로 변환
			
			Writer writer = new OutputStreamWriter(os, "UTF-8");
			
			writer.write(str);
			
			writer.flush();
			writer.close();
			
			
			
		} catch (FileNotFoundException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch(IOException ioe) {
			System.out.println(ioe.getMessage());
		} catch(Exception e) {
			System.out.println(e.getMessage());
		}
	}
	
	
	//InputStreamReader를 이용해서 
	//바이트스트림 -> 문자스트림으로 변환
	
	public static String read() {
		String returnStr = "";
		
		try {
			InputStream is = new FileInputStream("C:/java/hp01.txt");
			
			Reader reader = new InputStreamReader(is);
			
			char[] chArr = new char[4096];
			
			
			while(true) {
				int cnt = reader.read(chArr);
						
						if(cnt == -1) {
						break;
						}
						
						for(int i = 0; i < cnt; i++) {
							returnStr = returnStr + 
									String.valueOf(chArr[i]);
						}
						
			
			}
			
			
			
		} catch (FileNotFoundException fnfe) {
			// TODO Auto-generated catch block
			fnfe.printStackTrace();
		} catch(IOException ioe) {
			System.out.println(ioe.getMessage());
		} catch(Exception e) {
			System.out.println(e.getMessage());
		}
		return returnStr;
		
	}
	
	
	
	
	
	
	
	
	
	
	
}
